fury/actor.py

```diff
@@ -1440,42 +1446,92 @@ def dots(points, color=(1, 0, 0), opacity=1, dot_size=5):
-            cnt += 1
-    else:
-        polyVertexPoints.InsertPoint(cnt, points[0], points[1], points[2])
-        aPolyVertex.GetPointIds().SetId(cnt, cnt)
-        cnt += 1
+    if points.ndim != 2:
+        raise ValueError('Invalid points. The shape of the structure must be '
+                         '(Nx3). Your data has {} dimensions.'
+                         .format(points.ndim))
+
+    if points.shape[1] != 3:
+        raise ValueError('Invalid points. The shape of the last dimension '
+                         'must be 3. Your data has a last dimension of {}.'
+                         .format(points.shape[1]))
+
+    vtk_vertices = Points()
+    vtk_faces = CellArray()
+
+    # Add points
+    for i in range(len(points)):
+        p = points[i]
+        idd = vtk_vertices.InsertNextPoint(p)
+        vtk_faces.InsertNextCell(1)
+        vtk_faces.InsertCellPoint(idd)
+
+    color_tuple = __color_check(points, colors)
+    color_array, global_opacity = color_tuple
+
+    # Create a polydata object
+    polydata = PolyData()
+    polydata.SetPoints(vtk_vertices)
+    polydata.SetVerts(vtk_faces)
+    polydata.GetPointData().SetScalars(color_array)
+
+    # Visualize
+    mapper = PolyDataMapper()
+    mapper.SetInputData(polydata)
+
+    # Create an actor
+    poly_actor = Actor()
+    poly_actor.SetMapper(mapper)
 
-    aPolyVertexGrid = UnstructuredGrid()
-    aPolyVertexGrid.Allocate(1, 1)
-    aPolyVertexGrid.InsertNextCell(aPolyVertex.GetCellType(),
-                                   aPolyVertex.GetPointIds())
+    if opacity is not None:
+        poly_actor.GetProperty().SetOpacity(opacity)
+    elif global_opacity >= 0:
+        poly_actor.GetProperty().SetOpacity(global_opacity)
+    poly_actor.GetProperty().SetPointSize(dot_size)
 
-    aPolyVertexGrid.SetPoints(polyVertexPoints)
-    aPolyVertexMapper = DataSetMapper()
-    aPolyVertexMapper.SetInputData(aPolyVertexGrid)
-    aPolyVertexActor = Actor()
-    aPolyVertexActor.SetMapper(aPolyVertexMapper)
+    return poly_actor
+
+
+def __color_check(dots, colors=None):
+    """
+    Returns a VTK scalar array containing colors information for each one of
+    the dots according to the policy defined by the parameter colors.
+
+    Parameters
+    ----------
+    dots : (N, 3) array or ndarray
+        points coordinates array.
+    colors : None or tuple (3D or 4D) or array/ndarray (N, 3 or 4)
+        If None a predefined color is used for each dot.
+        If one tuple of color is used. Then all dots will have the same color.
+        If an array (N, 3 or 4) is given, where N is equal to the number of
+        points. Then every point is colored with a different RGB(A) color.
+
+    Returns
+    -------
+    color_array : vtkDataArray
+        vtk scalar array with name 'colors'.
+    global_opacity : float
+        returns 1 if the colors array doesn't contain opacity otherwise -1.
+
+    """
+    num_dots = len(dots)
```

Comment:
Since you only need the length of the array, you could just pass that as an integer.